File: c_ds/train/basics/decode_msg.cpp
#include <unordered_map>
#include <iostream>

using namespace std;

class Solution {
public:
    string decodeMessage(string key, string message) {
        std::unordered_map<char, int> keyMap;
        int key_decode_offset = 0;
        string charmap = "abcdefghijklmnopqrstuvwxyz";
        string result;
        result.resize(message.length());
        for (auto key_char : key){
            if ( key_char == ' ') continue;
            if ( keyMap.find(key_char) == keyMap.end()){
                keyMap[key_char] = int(charmap[key_decode_offset]);
                key_decode_offset ++;
            }
        }
        int decode_offset = 0;
        for (auto enchar: message){
            if (enchar == ' '){
                result[decode_offset]  = enchar;
                decode_offset ++;
            }else{
                result[decode_offset] = char(keyMap[enchar]);
                decode_offset ++;
            }
        }
        return result;
    }
};

void test(){
    Solution s;
    string key = "the quick brown fox jumps over the lazy dog";
    string message = "vkbs bs t suepuv";
    cout << "\t message:" << message << "\n\t decoded:" << s.decodeMessage(key, message) << endl;
}

int main(){
    test();
    return 0;
}

File: c_ds/train/basics/decode_msg.py
class Solution:
    def decodeMessage(self, key: str, message: str) -> str:
        decod_key = {}
        result = ""
        val_arr = 'abcdefghijklmnopqrstuvwxyz'
        key_offset = 0
        for key_char in key:
            if key_char == ' ':
                continue
            if key_char not in decod_key.keys():
                # print(f'{key_offset}:{key_char}')
                decod_key[key_char] = ord(val_arr[key_offset])
                key_offset += 1
        # print(decod_key)
        for en_char in message:
            if en_char == " ":
                result += en_char
            else:
                result += chr(decod_key[en_char])
        return result


def main():
    s = Solution()
    key = "the quick brown fox jumps over the lazy dog"
    message = "vkbs bs t suepuv"
    print(f"\t message:{message} \n\t decoded:{s.decodeMessage(key, message)}")


if __name__ == "__main__":
    main()